unittests/test-contracts/svnn_ibc/svnn_ibc.hpp

```diff
@@ -329,15 +331,16 @@ CONTRACT svnn_ibc : public contract {
 
          //returns hash of major_version + minor_version + finalizer_policy_generation + resolve_witness() + finality_mroot
          checksum256 finality_digest() const {

```

Comment:
What if we rearranged it to something like this? If you like this approach, I can use that pattern instead. 


      struct block_finality_data {
         
         //major_version for this block
         uint32_t major_version;

         //minor_version for this block
         uint32_t minor_version;

         //finalizer_policy_generation for this block
         uint32_t finalizer_policy_generation;

         uint32_t final_on_qc_block_num = 0;

         std::optional<finalizer_policy_input> new_finalizer_policy;

         //if a finalizer policy is present, witness_hash should be the base_digest. Otherwise, witness_hash should be the static_data_digest
         checksum256 witness_hash;

         //final_on_qc for this block
         checksum256 finality_mroot;
         
         //returns hash of digest of new_finalizer_policy + witness_hash if new_finalizer_policy is present, otherwise returns witness_hash
         checksum256 resolve_witness() const {
            if (new_finalizer_policy.has_value()){
               checksum256 policy_digest = new_finalizer_policy.value().digest();
               checksum256 base_fpolicy_digest = hash_pair( std::make_pair( policy_digest, witness_hash) );
               return base_fpolicy_digest;
            }
            else {
               return witness_hash;
            }
         }; 

      };

      struct block_finality_data_internal : block_finality_data {

         checksum256 witness_cache;

         block_finality_data_internal(const block_finality_data& base) : block_finality_data(base){
            witness_cache = base.resolve_witness();
         }

         checksum256 finality_digest() const {
            auto pack_size = eosio::pack_size(*this);
            auto result = eosio::pack(*this);
            checksum256 hash = sha256(reinterpret_cast<char*>(result.data()), pack_size);
            return hash;
         }

         EOSLIB_SERIALIZE(block_finality_data_internal, (major_version)(minor_version)(finalizer_policy_generation)(final_on_qc_block_num)(finality_mroot)(witness_cache))

      };

Reply:
@systemzax this looks good to me. 
Do we have to call `eosio::pack_size(*this)`, or could we just use `result.size()`?

Reply:
Also, why do we need `reinterpret_cast<char*>`? doesn't `pack()` return a `vector<char>`?

Reply:
I fixed both of these issues in my branch. I will refactor the rest of custom digest functions to use a similar pattern.

Reply:
Great, thanks @systemzax